wpilib/wpilib/watchdog.py

```diff
@@ -26,45 +30,137 @@ class Watchdog:
     The watchdog is initialized disabled, so the user needs to call enable() before use.
     """
 
-    def __init__(self, timeout: float, callback=lambda: None) -> None:
+    # Used for timeout print rate-limiting
+    kMinPrintPeriod = 1000000  # us
+
+    _watchdogs = []
+    _queueMutex = threading.Lock()
+    _schedulerWaiter = threading.Condition(_queueMutex)
+    _keepAlive = True
+    _thread = None
+
+    @classmethod
+    def _reset(cls) -> None:
+        thread = cls._thread
+        if thread is not None:
+            cls._keepAlive = False
+            with cls._queueMutex:
+                cls._watchdogs.clear()
+                cls._schedulerWaiter.notify_all()
+            thread.join()
+        cls._thread = None
+        cls._keepAlive = True
+
+    def __init__(self, timeout: float, callback: Callable[[], None]) -> None:
         """Watchdog constructor.
 
-        :param timeout: The watchdog's timeout in seconds.
+        :param timeout: The watchdog's timeout in seconds with microsecond resolution.
+        :param callback: This function is called when the timeout expires.
         """
-        self.timeout = timeout
-        self.callback = callback
-        self.startTime = 0.0
-        self.epochs = {}
+        self._startTime = 0  # us
+        self._timeout = int(timeout * 1e6)  # us
+        self._expirationTime = 0  # us
+        self._callback = callback
+        self._lastTimeoutPrintTime = 0  # us
+        self._lastEpochsPrintTime = 0  # us
+        self._epochs = {}  # type: Dict[str, int]
         self._isExpired = False
-        self.notifier = Notifier(self.timeoutFunc)
-        self.enable()
+
+        #: Enable or disable suppression of the generic timeout message.
+        #:
+        #: This may be desirable if the user-provided callback already
+        #: prints a more specific message.
+        self.suppressTimeoutMessage = False
+
+        Resource._add_global_resource(self)
```

Comment:
Hmm. If we have the `_reset` above, is this necessary? @virtuald

Reply:
I've gone ahead and removed this and the corresponding redundant `close` method then.